[YUNIKORN-1793] Handle placement rule and queue changes during initialisation

craigcondit opened the pull request:
### What is this PR for?
Adds a tag to mark applications as forced create. This allows most validation to be suppressed and if necessary, will redirect the app to a recovery queue.

### What type of PR is it?
* [ ] - Bug Fix
* [ ] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [x] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-1793

### How should this be tested?

### Screenshots (if appropriate)

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

pbacsko commented on pkg/scheduler/partition.go:
```diff
@@ -305,39 +305,85 @@ func (pc *PartitionContext) AddApplication(app *objects.Application) error {
 		queueName = app.GetQueuePath()
 		if queueName == "" {
-			return fmt.Errorf("application rejected by placement rules: %s", appID)
+			if forceCreate {
+				queueName = configs.RecoveryQueueFull
+				app.SetQueuePath(configs.RecoveryQueueFull)
+			} else {
+				return fmt.Errorf("application rejected by placement rules: %s", appID)
+			}
```
This part needs coverage.

craigcondit commented:
Updated approach to move the queue placement logic entirely within the placeholder manager and incorporated YUNIKORN-19 (always use placement manager). This simplifies the logic considerably as the placement manager is now always active and if it returns a queue, that queue is always to be created if it does not exist.

wilfred-s commented on pkg/scheduler/placement/provided_rule.go:
```diff
@@ -55,9 +56,10 @@ func (pr *providedRule) initialise(conf configs.PlacementRule) error {
 	// since this is the provided rule we must have a queue in the info already
 	queueName := app.GetQueuePath()
-	if queueName == "" {
+	if queueName == "" || common.IsRecoveryQueue(queueName) {
```
The queue name can have a prefix of `root.` or might not have it. I can even use provided with a parent rule to and make it really tricky to detect that it would end up in the recovery queue.
A simple ACL check on the recovery queue can prevent all that from being possible.

craigcondit replied:
This does seem cleaner. However, how do we enforce a static ACL on the recovery queue? The ACLs get populated via configuration, and since that config won't exist for the recovery queue, the ACLs will be empty.

craigcondit replied:
Fixed this via using an impossible ACL for the recovery queue ; all rules (except) recovery now return a flag indicating that queue ACLs need to be checked. The recovery rule returns false, allowing the normal ACL checks to be skipped and the recovery queue used.

wilfred-s reviewed (changes requested):
A 'recovery' rule (not available in the configurable list) pointing to a fixed queue (cannot be created in the config) which has a fixed ACL matching to an invalid user that cannot be specified all based on a tag that the k8shim always sets based on the state of init will I think simplify this solution.

craigcondit commented:
Rebased on latest master.

craigcondit commented on pkg/scheduler/objects/application.go:
```diff
@@ -1875,9 +1875,6 @@ func (sa *Application) GetUser() security.UserGroup {
 // Get a tag from the application
 // Note: tags are not case sensitive
 func (sa *Application) GetTag(tag string) string {
```
Tags are currently immutable in the application. Therefore, access is safe. You are correct that if this changes in the future race conditions could result. However, that I think is out of scope here.

craigcondit commented on pkg/scheduler/partition.go:
```diff
@@ -306,38 +297,41 @@ func (pc *PartitionContext) AddApplication(app *objects.Application) error {
-		if err != nil {
-			return fmt.Errorf("failed to create rule based queue %s for application %s", queueName, appID)
+		if common.IsRecoveryQueue(queueName) {
+			queue, err = pc.createRecoveryQueue()
```
It should not fail, however it's still possible (and is most definitely a bug if it does). Nothing configuration-related is checked, only sanity checks such as the root queue being provided properly.

wilfred-s commented on pkg/scheduler/objects/queue.go:
```diff
@@ -145,16 +147,32 @@ func NewConfiguredQueue(conf configs.QueueConfig, parent *Queue) (*Queue, error)
 
+// NewRecoveryQueue creates a recovery queue if it does not exist. The recovery queue
+// is a dynamic queue, but has an invalid name so that it cannot be directly referenced.
+func NewRecoveryQueue(parent *Queue) (*Queue, error) {
```
The config update should never accept a recovery queue. The queue should not be allowed in the config as it does not pass the queue name checks. That means if someone tries to create it in the config the update should be rejected outright and not cause an issue.

For the next part: the recovery queue should not get a child template applied. Applying the child template to the recovery queue could cause not just the sort order to be overridden but could also change quota or preemption. Not something we want to allow I think.

craigcondit replied:
Recovery queue refactored to set an impossible ACL and ignore template settings.

pbacsko reviewed (approved):
+1 LGTM (after Wilfred's comments are addressed)

craigcondit commented on pkg/common/security/usergroup.go:
```diff
@@ -122,14 +123,25 @@ func (c *UserGroupCache) resetCache() {
-		return UserGroup{}, fmt.Errorf("empty user cannot resolve")
+		if force {
+			// app creation is forced, so we need to synthesize a user / group
+			ugi.User = common.AnonymousUser
```
Discussed this offline and decided to keep it anonymous as this is not a recovery user at all - simply a user that is used in the case where recovery needs to happen and UGI lookup fails.
